### PR TITLE
Introduce Error binding patterns

### LINKED ISSUE #13889: Error Binding Pattern
**Description:**
Error binding pattern is defined in the spec as below:
```
error-binding-pattern :=
   error ( simple-binding-pattern [, error-detail-binding-pattern] )
error-detail-binding-pattern :=
   simple-binding-pattern | record-binding-pattern
```

- an error-binding-pattern error(rp, dp) matches an error value if rp matches the error reason and dp matches the error detail record;
- an error-binding-pattern error(rp) is equivalent to error(rp, _)
- Given a type descriptor for every variable in a binding-pattern, there is a type descriptor for the binding-pattern; for an error-binding-pattern, the type descriptor is an error type descriptor.

As per the spec, the following variable declarations can be written:
```ballerina
    error errorOne = error("Error One", { message: "Msg One", detail: "Detail Msg" });
    error<string, map<string|boolean>> errorTwo = error("Error Two", { message: "Msg Two", fatal: true });

    var error (reasonString, detailMap) = errorOne;
    error<string, map<string|boolean>> error (reasonString1, { message, fatal }) = errorTwo;
```
And the following variable reference patterns can be written:
```ballerina
    error errorOne = error("Error One", { message: "Msg One", detail: "Detail Msg" });
    error<string, map<string|boolean>> errorTwo = error("Error Two", { message: "Msg Two", fatal: true });

    string reasonString;
    map<any> detailMap;
    string reasonString1;
    string|boolean? message;
    string|boolean? fatal;

    error (reasonString, detailMap) = errorOne;
    error (reasonString1, { message, fatal }) = errorTwo;
```

The error binding pattern should also be supported for static match statment:
```
error-match-pattern :=
   error ( simple-match-pattern [, error-detail-match-pattern] )
error-detail-match-pattern :=
   simple-match-pattern | record-match-pattern
```

Following are the valid syntax for error binding pattern within the match statement:
```ballerina
function fooFunction(any|error e) returns string {
    match e {
        var (a, b) => return "Matched with tuple var : " + io:sprintf("%s", a);
        var { a, b } => return "Matched with record var : " + io:sprintf("%s", a);
        var error (reason, detail) => return "Matched with error var : " + io:sprintf("%s", reason);
        var x => return "Matched nothing : " + io:sprintf("%s", x);
    }
}
```

Sub task of #10525

### PR DESCRIPTION
**Pull request by Anoukh**
## Purpose
This PR introduces Error Binding Pattern which includes variable definition, variable reference and structured error match pattern
Closes #13889

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 1]
```diff
@@ -701,7 +705,24 @@ public void visit(BLangRecordVariableDef varDefNode) {
 
     @Override
     public void visit(BLangErrorVariableDef varDefNode) {
-        // TODO: complete
+        BLangErrorVariable errorVariable = varDefNode.errorVariable;
+
+        //create error destruct block stmt
```
Comment (Shan1024) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java]:
```suggestion
        // Create error destruct block stmt.
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 2]
```diff
@@ -701,7 +705,24 @@ public void visit(BLangRecordVariableDef varDefNode) {
...
+        final BLangBlockStmt blockStmt = ASTBuilderUtil.createBlockStmt(varDefNode.pos);
+
+        //create a simple var for the error 'error x = ($error$)'
```
Comment (Shan1024) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java]:
```suggestion
        // Create a simple var for the error 'error x = ($error$)'.
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 3]
```diff
@@ -701,7 +705,24 @@ public void visit(BLangRecordVariableDef varDefNode) {
...
+        final BLangSimpleVariable error = ASTBuilderUtil.createVariable(varDefNode.pos, "", symTable.errorType, null,
+                new BVarSymbol(0, names.fromString("$error$"), this.env.scope.owner.pkgID, symTable.errorType,
+                        this.env.scope.owner));
+        error.expr = errorVariable.expr;
+        final BLangSimpleVariableDef variableDef = ASTBuilderUtil.createVariableDefStmt(varDefNode.pos, blockStmt);
+        variableDef.var = error;
+
+        //create the variable definition statements using the root block stmt created
```
Comment (Shan1024) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java]:
```suggestion
        // Create the variable definition statements using the root block stmt created.
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 4]
```diff
@@ -701,7 +705,24 @@ public void visit(BLangRecordVariableDef varDefNode) {
...
+        createVarDefStmts(errorVariable, blockStmt, error.symbol, null);
+
+        //finally rewrite the populated block statement
```
Comment (Shan1024) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java]:
```suggestion
        // Finally rewrite the populated block statement.
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 5]
```diff
@@ -722,31 +743,43 @@ private void createVarDefStmts(BLangTupleVariable parentTupleVariable, BLangBloc
...
+
             if (NodeKind.VARIABLE == variable.getKind()) { //if this is simple var, then create a simple var def stmt
-                BLangLiteral indexExpr = ASTBuilderUtil.createLiteral(variable.pos, symTable.intType, (long) index);
                 createSimpleVarDefStmt((BLangSimpleVariable) variable, parentBlockStmt, indexExpr, tupleVarSymbol,
                         parentIndexAccessExpr);
-            } else if (variable.getKind() == NodeKind.TUPLE_VARIABLE) { //else recursively create the var def statements
+                continue;
+            }
+
+            if (variable.getKind() == NodeKind.TUPLE_VARIABLE) { //else recursively create the var def statements
```
Comment (Shan1024) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java]:
```suggestion
            if (variable.getKind() == NodeKind.TUPLE_VARIABLE) { // Else recursively create the var def statements.
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 6]
```diff
@@ -1150,6 +1285,22 @@ private void createVarRefAssignmentStmts(BLangTupleVarRef parentTupleVariable, B
                 }
                 createVarRefAssignmentStmts((BLangRecordVarRef) expression, parentBlockStmt, tupleVarSymbol,
                         arrayAccessExpr);
+                continue;
+            }
+
+            if (expression.getKind() == NodeKind.ERROR_VARIABLE_REF) {
+                //else recursively create the var def statements for record var ref
```
Comment (Shan1024) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java]:
```suggestion
                // Else recursively create the var def statements for record var ref.
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java [hunk 7]
```diff
@@ -740,14 +741,16 @@ private boolean checkStructuredPatternSimilarity(BLangVariable precedingVar, BLa
 
             for (int i = 0; i < precedingRecVar.variableList.size(); i++) {
                 BLangRecordVariableKeyValue precedingKeyValue = precedingRecVar.variableList.get(i);
-                if (!recVarAsMap.containsKey(precedingKeyValue.key.value)) {
-                    return false;
+                if (recVarAsMap.containsKey(precedingKeyValue.key.value)) {
```
Comment (Shan1024) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java]: I think it is better to inverse the `if` here to reduce the number of nested statements.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java [hunk 8]
```diff
@@ -535,12 +537,27 @@ public void visit(BLangTupleVariable varNode) {
...
+        if (varNode.type == null) {
+            varNode.type = symResolver.resolveTypeNode(varNode.typeNode, env);
+        }
+        if (!validateErrorVariable(varNode)) {
+            varNode.type = symTable.semanticError;
+            return;
+        }
+        symbolEnter.defineNode(varNode, env);
+        if (varNode.expr == null) {
+            // we have no rhs to do type checking
```
Comment (Shan1024) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java]:
```suggestion
            // We have no rhs to do type checking.
```

**Comment by hasithaa** [comment]
@Anoukh There are compilation errors. Can you fix those?

**Pull request merged**